kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -117,307 +134,120 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
+            val curOwner = this.owner.value
+            if (curOwner === NO_OWNER) continue // <-- ATTENTION, BLOCKING PART HERE
+            // Check the owner.
+            check(curOwner === owner) { "This mutex is locked by $curOwner, but $owner is expected" }
+            // Try to clean the owner first. We need to use CAS here to synchronize with concurrent `unlock(..)`-s.
+            if (!this.owner.compareAndSet(owner, NO_OWNER)) continue
+            // Release the semaphore permit at the end.
+            release()
+            return
         }
     }
 
-    override fun toString(): String {
-        _state.loop { state ->
-            when (state) {
-                is Empty -> return "Mutex[${state.locked}]"
-                is OpDescriptor -> state.perform(this)
-                is LockedQueue -> return "Mutex[${state.owner}]"
-                else -> error("Illegal state $state")
-            }
-        }
-    }
+    @Suppress("UNCHECKED_CAST")
+    override val onLock: SelectClause2<Any?, Mutex> get() = SelectClause2Impl(
+        clauseObject = this,
+        regFunc = MutexImpl::onLockRegFunction as RegistrationFunction,
+        processResFunc = MutexImpl::onLockProcessResult as ProcessResultFunction,
+        onCancellationConstructor = onCancellationUnlockConstructor
+    )
 
-    private class LockedQueue(
-        @JvmField var owner: Any
-    ) : LockFreeLinkedListHead() {
-        override fun toString(): String = "LockedQueue[$owner]"
+    protected open fun onLockRegFunction(select: SelectInstance<*>, owner: Any?) {
+        onAcquire.regFunc(this, SelectInstanceWithOwner(select, owner), owner)
     }
 
-    private abstract inner class LockWaiter(
-        @JvmField val owner: Any?
-    ) : LockFreeLinkedListNode(), DisposableHandle {
-        private val isTaken = atomic(false)
-        fun take(): Boolean = isTaken.compareAndSet(false, true)
-        final override fun dispose() { remove() }
-        abstract fun tryResumeLockWaiter(): Boolean
-        abstract fun completeResumeLockWaiter()
+    protected open fun onLockProcessResult(owner: Any?, result: Any?): Any? {
+        onAcquire.processResFunc(this, null, result)
+        return this
     }
 
-    private inner class LockCont(
-        owner: Any?,
-        private val cont: CancellableContinuation<Unit>
-    ) : LockWaiter(owner) {
+    protected val onCancellationUnlockConstructor: OnCancellationConstructor =
+        { select: SelectInstance<*>, owner: Any?, ignoredResult: Any? ->
+            { unlock(owner) }
+        }
 
-        override fun tryResumeLockWaiter(): Boolean {
-            if (!take()) return false
-            return cont.tryResume(Unit, idempotent = null) {
-                // if this continuation gets cancelled during dispatch to the caller, then release the lock
-                unlock(owner)
-            } != null
+    private inner class CancellableContinuationWithOwner(
+        val cont: CancellableContinuation<Unit>,
```

Comment:
Here and below please mark these `val`-s with `@JvmField` to save a few bytes in dex/bytecode/classfile size